[RFC, ready] Batched Inverse

vishwakftw opened the pull request:
Complete billing of changes:

Related to Batch Inverse:
- [x] Add batched inverse (CPU)
- [x] Add batched inverse (CUDA)
- [x] Modify autograd entry
- [x] Add tests
  - [x] test_autograd
  - [x] test_cuda
  - [x] test_torch
- [x] Modify docs
- [x] Remove `_batch_inverse` in `MultivariateNormal`.
- [x] Allow batch matrices as inputs for negative powers in `matrix_power`

Miscellaneous modifications:
- [x] Move all batch operations to BatchLinearAlgebra.cpp/.cu and provide general framework for adding more batch ops.
- [x] Add a RAII structure for MAGMA queue management.

vishwakftw committed:
Add batched inverse

vishwakftw committed:
Remove pin_memory from Gesv.cu

vishwakftw committed:
Missed #endif

vishwakftw committed:
Fix computation of inverses

I didn't know how getri worked after using getrf, which led to some issues

vishwakftw committed:
Silly typographical error; my bad

vadimkantorov commented:
For a test, e.g. this style transfer code https://github.com/NVIDIA/FastPhotoStyle/blob/master/photo_smooth.py#L77 batch-inverts matrices of shape HxWx3x3 in NumPy, one can compare perf of batched GPU inversion versus NumPy.

vishwakftw committed:
Modify docs, derivative, and add test in test_autograd

vishwakftw committed:
Add tests in test_cuda, test_torch

vishwakftw committed:
Remove _batch_inverse from MVN

vishwakftw committed:
Fix include error

vishwakftw commented:
Oops, I didn't notice https://github.com/pytorch/pytorch/pull/9102 . Should I close this?

vishwakftw committed:
Rename files for consistency

vishwakftw reviewed (commented):
After building, the times between the CPU version and the NumPy version looked similar, but there was a big performance regression in the CUDA case. Help will be appreciated, and I have added my own comments for improving down below.

vishwakftw committed:
Clean up error messages, pass infos by ref

vishwakftw committed:
Modify test case for full rank matrices

vishwakftw commented on test/test_autograd.py:
```diff
@@ -2537,11 +2537,21 @@ def make_nonzero_det(A, sign=None, min_singular_value=0.1):
-    u, _, v = A.svd()
-    s = torch.arange(1., l + 1).mul_(1.0 / (l + 1))
-    return u.mm(torch.diag(s)).mm(v.t())
+def random_fullrank_matrix_distinct_singular_value(l, batches=None):
```
I modified the function to generate a batch of random full rank matrices.

vishwakftw replied:
However despite this, the CUDA tests are failing.

vishwakftw committed:
Some changes

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw commented:
The performance regression was for small matrices (5 x 5). When I tried benchmarking CUDA batch inverse of batch size 2 x 3 with individual matrix size of 100 x 100, there was a 10x speedup.

vishwakftw committed:
exclude gesv and inverse functions from test_jit

vishwakftw committed:
fix nit

vadimkantorov commented:
What about 3x3 or 4x4? (frequent in geometry applications, though in practice one needs a batches linear solver instead)

vishwakftw committed:
Remove unnecessary Python bindings, fix inverse for empty tensors

vishwakftw commented:
@vadimkantorov As mentioned above, CUDA doesn't work fast enough for small matrices. This is the context of https://github.com/pytorch/pytorch/issues/10172 .

vishwakftw commented:
@zou3519 Could I get a review?

zou3519 commented:
@vishwakftw I'll add this to my queue, will try to get to it soon. In the meantime could you rebase this so we can see the test results?

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
fix CUDA build failure

yf225 commented:
@vishwakftw Do you mind rebasing it again? Also ping @zou3519 for review :)

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
Update MiscUtils.h

vishwakftw committed:
Fix lint

vishwakftw committed:
Fix CUDA build failure

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
Fix some nits

vishwakftw committed:
Merge branch 'batch-inverse' of github.com:vishwakftw/pytorch into batch-inverse

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw commented on aten/src/ATen/native/cuda/Inverse.cu:
```diff
@@ -0,0 +1,156 @@
+
+  magmaGetriBatched<scalar_t>(
+    n, self_array, n, ipiv_array, self_inv_array,
+    n, info_array, batch_size, inverse_magma_queue);
```
Is reusing the same queue causing errors in the tests?

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
Remove check after batched getrf

vishwakftw committed:
self.type()._inverse_helper(self) --> at::_inverse_helper(self)

t-vi commented:
What's the perspective for having Magma-functions in ATen native? I realize that there is quite a bit of code, but having each (ATen visible) function in its own file seems like a lot of clutter if we eventually do this for all of THCTensorMathMagma.

vishwakftw commented:
@t-vi Have a look at https://github.com/pytorch/pytorch/issues/11325 .

t-vi commented:
No, I don't mean the functions, but the files. Having `native/Gesv.h`, `native/Gesv.cpp`, `native/Gesv.cu` doesn't scale look pretty when multiplied by to x linear algebra functions.

vishwakftw commented:
I think we probably have to move them to a folder: `native/linalg` and `native/linalg/cuda` when the time comes. Yes, I do agree that the place would become cluttered.

t-vi commented:
How about having more of them in one file?

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-inverse

vishwakftw committed:
Refactor batch linear operations

This fixes the issue raised by @t-vi regarding the possibility of a whole
bunch of files piling up in aten/native

Effectively, what this commit does is as follows
- move the utilities associated with batch operations to LinearAlgebraUtils.h
- make the process of adding a batch linalg function generic, with minimal effort.
  Future contributors will have to deal with BatchLinearAlgebra.cpp / .cu to add
  more functions

vishwakftw committed:
Revert submodule change, and revert changes to test_jit

vishwakftw committed:
Reorganize functionally

t-vi commented:
Hi,

would you be able to rebase this?
If not, I'll take the reorg bits in the batch cholesky PR so I'm independent.

Best regards

Thomas

vishwakftw commented:
I'll get it done right away.

vishwakftw commented:
The rebase got botched. Will do it again.

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-inverse

vishwakftw committed:
Fix one more conflict

vishwakftw committed:
add _inverse_helper to aten_interned_strings.h

vishwakftw committed:
Fix test error

vishwakftw committed:
Fix one more test

vishwakftw commented:
@zou3519 this is ready for review.

zou3519 commented:
@vishwakftw I'll try to get to it by monday. Please ping me again here or on slack if I don't

vishwakftw commented:
@zou3519 ping!

zou3519 commented on test/test_torch.py:
```diff
@@ -4629,24 +4629,66 @@ def test(name):
-        M = torch.randn(5, 5)
+    @staticmethod
+    def _test_inverse(self, cast):
+        from common import random_fullrank_matrix_distinct_singular_value
```
did not know we had this...

zou3519 commented on test/test_torch.py:
```diff
@@ -4629,24 +4629,66 @@ def test(name):
+        from common import random_fullrank_matrix_distinct_singular_value
+
+        # no batches: 2-D tensors
+        M = cast(random_fullrank_matrix_distinct_singular_value(5))
```
nit: while we're touching this test anyways, it would be nice if these were not one-letter names "matrix, matrix_inverse, expected" would be great

zou3519 commented on test/test_torch.py:
```diff
@@ -4239,24 +4239,64 @@ def test(name):
+        M = cast(torch.randn(5, 5))
         MI = torch.inverse(M)
-        E = torch.eye(5)
-        self.assertFalse(MI.is_contiguous(), 'MI is contiguous')
```
I'm confused: shouldn't the output of torch.inverse be non-contiguous?

zou3519 commented on test/test_torch.py:
```diff
@@ -4629,24 +4629,66 @@ def test(name):
-    def test_inverse(self):
-        M = torch.randn(5, 5)
+    @staticmethod
+    def _test_inverse(self, cast):
```
nit: It's nicer to pass device instead of cast

zou3519 commented on test/test_torch.py:
```diff
@@ -4629,24 +4629,66 @@ def test(name):
+        MI = torch.Tensor(3, 5, 5)
+        torch.inverse(M, out=MI)
+        self.assertEqual(torch.inverse(M), MI)
+
```
We only test `torch.inverse` on tensors of the form `(B, N, N)` and `(N, N)`. Would it be possible to add one test for a tensor of the form `(B1, ..., BK, N, N)`?

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -1987,15 +1987,18 @@ def parse_kwargs(desc):
-Takes the inverse of the square matrix :attr:`input`.
+Takes the inverse of the square matrix :attr:`input`. :attr:`input` can be batches
+of 2D square tensors, in which case this function would return a tensor composed of
+individual inverses.
```
What happens if the inverse can't be computed? Just wondering

vishwakftw replied:
An error will be thrown. We have checks in internal code to take care of this.

zou3519 commented on test/test_autograd.py:
```diff
@@ -3132,7 +3132,9 @@ class dont_convert(tuple):
-    ('inverse', (S, S), NO_ARGS, '', NO_ARGS, [skipIfNoLapack]),
+    ('inverse', random_fullrank_matrix_distinct_singular_value(S), NO_ARGS, '', NO_ARGS, [skipIfNoLapack]),
+    ('inverse', random_fullrank_matrix_distinct_singular_value(S, 2, 3),
+     NO_ARGS, 'batched', NO_ARGS, [skipIfNoLapack]),
```
Would be nice to add a test case for a tensor of size `(B1, B2, B3, N, N)` (more than one batch dimension)

vishwakftw replied:
This is already for more than one batch dimensions: the shape tested here is `2 x 3 x S x S`.

zou3519 replied:
My bad, I didn't realize

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
CR

vishwakftw committed:
Fix nits

vishwakftw committed:
Fix more nits

sorry.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -0,0 +1,209 @@
+    return at::_getri_single(self);
+  }
+  inverseCheckInputs(self);
+  return at::_inverse_helper(self);
```
Hmm would prefer this be more explicit that it was generated from LINALG_HELPER_1_ARGS.
Could we name the function `::_generate_inverse_linalg_helper` and the macro `GENERATE_LINALG_HELPER_1_ARGS` or something like that?

zou3519 replied:
Oh, I see. _inverse_helper isn't generated; _inverse_helper_cuda / inverse_helper_cpu are. Could we rename those so that it is clearer that they were generated?

vishwakftw replied:
Thanks. I'll rename the macros to `GENERATE_LINALG_HELPER_n_ARGS`.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -0,0 +1,209 @@
+  AT_ERROR("inverse: LAPACK library not found in compilation");
+#endif
+  auto self_data = self.data<scalar_t>();
+  auto self_matrix_stride = matrixStride(self);
```
Maybe add an assert that self has memory in column major form? (that is the case, right?)

zou3519 commented:
Tests are failing

vishwakftw committed:
Fix tests and CR

vishwakftw committed:
Merge branch 'master' into batch-inverse

vishwakftw committed:
Merge branch 'batch-inverse' of github.com:vishwakftw/pytorch into batch-inverse

vishwakftw committed:
Fix tests, build
- matrix_power can now take batches of matrices for negative powers

vishwakftw committed:
Undo submodule change, modify docs for matrix_power

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-inverse

zou3519 commented on aten/src/ATen/native/cuda/MiscUtils.h:
```diff
@@ -0,0 +1,50 @@
+namespace native {
+
+#ifdef USE_MAGMA
+static magma_queue_t createMagmaQueue(const Tensor& tensor) {
```
We should make this RAII given that we always want to delete the created queue

vishwakftw replied:
We can make this RAII if there is no considerable overhead associated with creating a queue (RAII will mandate the creation of two queues, one for `getri` and one for `getrf`).

zou3519 replied:
Sorry, I'm not following: the magma queue seems local to the function, so wouldn't getri and getrf use different queues regardless of if magma queue creation is RAII or not?

To clarify with gesv as an example, the code currently looks like this:
```
  magma_queue_t gesv_magma_queue = createMagmaQueue(b);
   magmaGesvBatched<scalar_t>(
      n, nrhs, A_array, n, ipiv_array, b_array, n,
      info_array, batch_size, gesv_magma_queue);
   destroyMagmaQueue(gesv_magma_queue);
```
but we could prevent future contributors from accidentally forgetting to free the queue with something like the following:
```
  MagmaQueue gesv_magma_queue;
   magmaGesvBatched<scalar_t>(
      n, nrhs, A_array, n, ipiv_array, b_array, n,
      info_array, batch_size, gesv_magma_queue);
  // no need to destory gesv_magma_queue, it will be cleaned up when it goes out of scope
```

vishwakftw replied:
I seemed to have misunderstood your pitch earlier, and I think understand now. Thank you for clarifying.

In my earlier comment I thought of something like this:
```
   magmaGesvBatched<scalar_t>(
      n, nrhs, A_array, n, ipiv_array, b_array, n,
      info_array, batch_size, createMagmaQueue(b)); 
// This would be cleaned up, right? - since it is local to the function.
```

zou3519 replied:
Oh I see. Yes, in that case, The magma queue would only live inside the call to `magmaGesvBatched`

vishwakftw replied:
I will change the functions to do this instead, and remove `destroyMagmaQueue`. Thanks for the suggestion.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -0,0 +1,216 @@
+}
+
+// Because this is out-of-place inverse, the predefined macros will
+// not work
```
Maybe we can define a macro for this in the future if we add more out-of-place lapack functions...

vishwakftw replied:
Sure, I will check if there are out-of-place MAGMA functions in addition to `getri`.

zou3519 reviewed (approved):
One minor comment, otherwise, lgtm!

vishwakftw commented:
There is an optimization available for CUDA batched `getrf` for square matrices of dim < 32. Should I include this optimization here, or leave it for a later PR?

zou3519 commented:
@vishwakftw if it's a quick change feel free to do it here. Otherwise, it'll be easy to review as a separate PR

vishwakftw committed:
CR and improvements
- use magma_*getrf_smallsq_shfl for batches of matrices with dim <= 32
- remove destroyMagmaQueue and make magma queue more RAII-like

vishwakftw commented on aten/src/ATen/native/cuda/MiscUtils.h:
```diff
@@ -22,15 +22,11 @@ static magma_queue_t createMagmaQueue(const Tensor& tensor) {
   return magma_queue;
 }
 
-static void destroyMagmaQueue(magma_queue_t& existing_queue) {
```
@zou3519 Trust this is okay.

vishwakftw commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -179,17 +199,19 @@ AT_ERROR("inverse: MAGMA library not found in "
+  if (self.size(-2) <= 32) {
+    magmaGetrfSmallSquareBatched<scalar_t>(
+      n, self_array, n, ipiv_array, info_array,
+      batch_size, createMagmaQueue(self));
```
And this too.

vishwakftw commented:
Seemingly, the optimization for small matrix getrf is done inside MAGMA's getrf functions itself. I found this while inspecting the source code. I'll revert that change.

vishwakftw committed:
Add RAII struct for MAGMA queue management, revert getrf optimization

vishwakftw commented on aten/src/ATen/native/cuda/MiscUtils.h:
```diff
@@ -0,0 +1,64 @@
+#ifdef USE_MAGMA
+
+// RAII for a MAGMA Queue
+struct MAGMAQueue {
```
@zou3519 I've added the RAII structure. Please let me know if this is what you had in mind.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -0,0 +1,208 @@
+
+  magmaGetriBatched<scalar_t>(
+    n, self_array, n, ipiv_array, self_inv_array,
+    n, info_array, batch_size, MAGMAQueue(self.get_device()));
```
You should share this MAGMAQueue between getrf and getri. Something like the following:

```
MAGAMAQueue magma_queue(self.get_device());

  magmaGetrfBatched<scalar_t>(
    n, n, self_array, n, ipiv_array, info_array,
    batch_size, magma_queue);
   magmaGetriBatched<scalar_t>(
    n, self_array, n, ipiv_array, self_inv_array,
    n, info_array, batch_size, magma_queue);

```

vishwakftw replied:
Sure, thank you.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -0,0 +1,208 @@
+void magmaGesvBatched(
+    magma_int_t n, magma_int_t nrhs, scalar_t** dA_array, magma_int_t ldda,
+    magma_int_t** dipiv_array, scalar_t** dB_array, magma_int_t lddb,
+    magma_int_t* dinfo_array, magma_int_t batch_count, MAGMAQueue magma_queue) {
```
These should be pass by const reference now (`const MAGMAQueue&`) so we don't copy the entire data structure when it is passed into these functions

vishwakftw replied:
Understood, my bad. I'll fix it.

vishwakftw committed:
CR

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-inverse

zou3519 commented on aten/src/ATen/native/cuda/MiscUtils.h:
```diff
@@ -0,0 +1,64 @@
+struct MAGMAQueue {
+
+  // Default constructor, does nothing.
+  MAGMAQueue() = default;
```
Er maybe let's delete this. `MagmaQueue() = delete`;
It doesn't make sense for people to create a MAGMAQueue via the default constructor because then magma_queue_destroy will be called on an uninitialized magma_queue_t in the destructor.

vishwakftw replied:
Done! Thank you for the suggestion.

zou3519 reviewed (approved):
Looks great now! One last comment :) then we can merge this.

The test failures are unrelated.

vishwakftw committed:
Delete default constructor for MAGMAQueue

Jasonhunger commented:
DTD_inv = torch.inverse(DTD + self.lambda3 * torch.eye(self.input_dim).cuda()) 
RuntimeError: inverse_cuda: For batch 0: U(704,704) is zero, singular U.
